FWCore/Modules/src/PathStatusFilter.cc

```diff
@@ -0,0 +1,434 @@
+#include "FWCore/Utilities/interface/Exception.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+
+#include <boost/spirit/include/phoenix_bind.hpp>
+#include <boost/spirit/include/qi.hpp>
+
+#include <memory>
+#include <ostream>
+#include <string>
+#include <sstream>
+#include <vector>
+#include <utility>
+
+namespace qi = boost::spirit::qi;
+namespace ascii = boost::spirit::ascii;
+namespace phoenix = boost::phoenix;
+
+namespace edm {
+
+  class EventSetup;
+  class StreamID;
+
+  namespace pathStatusExpression {
+    class Evaluator;
+  }
+
+  class PathStatusFilter : public global::EDFilter<> {
+  public:
+    explicit PathStatusFilter(ParameterSet const&);
+    static void fillDescriptions(ConfigurationDescriptions&);
+    virtual bool filter(StreamID, Event&, EventSetup const&) const override final;
+
+  private:
+    std::unique_ptr<pathStatusExpression::Evaluator> evaluator_;
+    bool verbose_;
+  };
+
+  namespace pathStatusExpression {
+
+    class Evaluator {
+    public:
+      virtual ~Evaluator() { }
+
+      enum EvaluatorType { Name, Not, And, Or, BeginParen };
+      virtual EvaluatorType type() const = 0;
+
+      virtual void setLeft(std::unique_ptr<Evaluator> &&) {}
+      virtual void setRight(std::unique_ptr<Evaluator> &&) {}
+
+      virtual void print(std::ostream & out, unsigned int indentation) const {}
+      virtual void init(ConsumesCollector &) {}
+      virtual bool evaluate(Event const& event) const { return true; };
+    };
+
+    class Operand : public Evaluator {
+    public:
+      Operand(std::vector<char> const& pathName) :
+        pathName_(pathName.begin(), pathName.end()) {
+      }
+
+      EvaluatorType type() const { return Name; }
+
+      void print(std::ostream & out, unsigned int indentation) const override {
+        out << std::string( indentation, ' ' ) << pathName_ << "\n";
+      }
+
+      void init(ConsumesCollector & iC) override {
+        token_ = iC.consumes<PathStatus>(InputTag(pathName_));
+      }
+
+      bool evaluate(Event const& event) const override {
+        Handle<PathStatus> handle;
+        event.getByToken(token_, handle);
+        return handle->accept();
+      }
+
+    private:
+      std::string pathName_;
+      EDGetTokenT<PathStatus> token_;
+    };
+
+    class NotOperator : public Evaluator {
+    public:
+      EvaluatorType type() const { return Not; }
+
+      virtual void setLeft(std::unique_ptr<Evaluator> && v) { operand_ = std::move(v); }
+
+      void print(std::ostream & out, unsigned int indentation) const override {
+        out << std::string( indentation, ' ' ) << "not\n";
+        operand_->print(out, indentation + 4);
+      }
+
+      void init(ConsumesCollector & iC) override {
+        operand_->init(iC);
+      }
+
+      bool evaluate(Event const& event) const override {
+        return !operand_->evaluate(event);
+      }
+
+    private:
+      std::unique_ptr<Evaluator> operand_;
+    };
+
+    class AndOperator : public Evaluator {
```

Comment:
You could templatize the binary operator so you can share code between `And` and `Or`.